src/osd/PrimaryLogPG.cc

```diff
@@ -9976,6 +9978,264 @@ struct C_Flush : public Context {
   }
 };
 
+struct C_SetDedupChunks : public Context {
+  PrimaryLogPGRef pg;
+  hobject_t oid;
+  epoch_t last_peering_reset;
+  ceph_tid_t tid;
+  uint64_t offset;
+  
+  C_SetDedupChunks(PrimaryLogPG *p, hobject_t o, epoch_t lpr)
+    : pg(p), oid(o), last_peering_reset(lpr),
+      tid(0)
+  {}
+  void finish(int r) override {
+    if (r == -ECANCELED)
+      return;
+    std::scoped_lock locker{*pg};
+    auto it = pg->manifest_ops.find(oid);
+    if (it == pg->manifest_ops.end()) {
+      // raced with cancel_manifest_ops
+      return;
+    }
+    // check if the previous op returns fail
+    if (it->second->results[0] < 0) {
+      return;
+    }
+
+    it->second->results[offset] = r;
+    if (last_peering_reset == pg->get_last_peering_reset() &&
+	r >= 0) {
+      if (it->second->num_chunks != it->second->results.size()) {
+	// there are on-going works
+	return;
+      }
+      pg->finish_set_dedup(oid, it->second->chunks, r, tid);
+    } else {
+      // if any failure occurs, put a mark on the results to recognize the failure
+      it->second->results[0] = r;
+      if (last_peering_reset != pg->get_last_peering_reset()) 
+	it->second->results[0] = -EINVAL;
+    }
+    pg->manifest_ops.erase(it);
+  }
+};
+
+
+int PrimaryLogPG::start_dedup(OpRequestRef op, ObjectContextRef obc)
+{
+  bufferlist bl;
+  uint64_t cur_off = 0;
+  const object_info_t& oi = obc->obs.oi;
+  const hobject_t& soid = oi.soid;
+
+  if (!obc->is_blocked()) {
+    return -EINVAL;
```

Comment:
I think you need to add more detail here.  If you want to disallow non-blocking writes for dedup objects, I think you probably want to reject the request much earlier -- probably in do_osd_ops.

Reply:
Take a look at how existing non-blocking flushes work.

Reply:
Probably in start_flush, if (!blocking) we should return an error.